### PR TITLE
Implement `ClientInformation` from spec

### LINKED ISSUE #1530: Implement API for fetching Agent identifier
Implement the solution to https://github.com/ably/specification/issues/111.

### PR DESCRIPTION
**Pull request by lawrence-forooghian**
Closes #1530. See commit messages for more details.

### CODE REVIEW

File: Source/ARTClientInformation.m [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+// NSOperatingSystemVersion has NSInteger as version components for some reason, so mitigate it here.
+static inline UInt32 conformVersionComponent(const NSInteger component) {
+    return (component < 0) ? 0 : (UInt32)component;
+}
+
+@implementation ARTClientInformation
+
++ (NSDictionary<NSString *, NSString *> *)agents {
+    NSMutableDictionary<NSString *, NSString *> *result = [NSMutableDictionary dictionary];
```
Comment (QuintinWillison) [inline Source/ARTClientInformation.m:19]: Pointer does not need to change:

```suggestion
    NSMutableDictionary<NSString *, NSString *> *const result = [NSMutableDictionary dictionary];
```

File: Source/ARTClientInformation.m [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    
+    [result addEntriesFromDictionary:[self platformAgent]];
+    [result addEntriesFromDictionary:[self libraryAgent]];
+
+    return result;
+}
+
++ (NSString *)agentIdentifierWithAdditionalAgents:(nullable NSDictionary<NSString *, NSString *> *)additionalAgents {
+    NSMutableDictionary<NSString *, NSString *> *agents = [self.agents mutableCopy];
```
Comment (QuintinWillison) [inline Source/ARTClientInformation.m:28]:
```suggestion
    NSMutableDictionary<NSString *, NSString *> *const agents = [self.agents mutableCopy];
```

File: Source/ARTClientInformation.m [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    
+    for (NSString *const additionalAgentName in additionalAgents) {
+        agents[additionalAgentName] = additionalAgents[additionalAgentName];
+    }
+    
+    return [self agentIdentifierForAgents:agents];
+}
+
++ (NSString *)agentIdentifierForAgents:(NSDictionary<NSString *, NSString*> *)agents {
+    NSMutableArray<NSString *> *components = [NSMutableArray array];
```
Comment (QuintinWillison) [inline Source/ARTClientInformation.m:38]:
```suggestion
    NSMutableArray<NSString *> *const components = [NSMutableArray array];
```

File: Source/ARTClientInformation.m [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
++ (NSDictionary<NSString *, NSString *> *)agents {
+    NSMutableDictionary<NSString *, NSString *> *result = [NSMutableDictionary dictionary];
+    
+    [result addEntriesFromDictionary:[self platformAgent]];
+    [result addEntriesFromDictionary:[self libraryAgent]];
+
+    return result;
+}
+
++ (NSString *)agentIdentifierWithAdditionalAgents:(nullable NSDictionary<NSString *, NSString *> *)additionalAgents {
```
Comment (QuintinWillison) [inline Source/ARTClientInformation.m:27]:
```suggestion
+ (NSString *)agentIdentifierWithAdditionalAgents:(nullable NSDictionary<NSString *, NSString *> *const)additionalAgents {
```

**Review by QuintinWillison — APPROVED** [review]
Good to go as is, but might be nice to patch a few opportunities to make intended pointer use self documenting...

**Review by maratal — APPROVED** [review]
LGTM 👍

**Pull request merged**